pyccel/parser/semantic.py

```diff
@@ -1570,6 +1570,14 @@ def _assign_lhs_variable(self, lhs, d_var, rhs, new_expressions, is_augassign =
                                         shape=a.alloc_shape, status=status))
                             args = new_args
                             new_args = []
+                    elif isinstance(lhs.class_type, (HomogeneousListType, HomogeneousSetType,DictType)):
+                        if isinstance(rhs, (PythonList, PythonDict, PythonSet)):
+                            alloc_type = 'init'
+                        elif isinstance(rhs, IndexedElement) or rhs.get_attribute_nodes(IndexedElement):
+                            alloc_type = 'resize'
+                        else:
+                            alloc_type = 'reserve'
+                        new_expressions.append(Allocate(lhs, shape=lhs.alloc_shape, status=status, alloc_type=alloc_type))
```

Comment:
I am concerned that this may break when we can return lists/sets/dicts from functions. What would be the correct case for this? `init`?

Reply:
done!